README
Note: For CERN testing...
Actual README for CERN is in CERNtestscripts directory





Overview of BSD CDAQ development software and hardware.
First Version Generated By: Tyler Anderson, Penn State University 
Date: Wed Mar 20 16:25:21 EDT 2013

For Questions and comments contact: 
Tyler Anderson (tba109@psu.edu) 
Sai Im (imsain@gmail.com)

-----------------------------------------------------------------------------------------
1.) BSD CDAQ development package overview

Parts in the hardware package:
a.) A simple USB board based around the FX2LP USB-2.0 microcontroller has been provided
b.) A power cable with flying leads for providing power to the USB board
c.) A simple jumper cable.

Instructions for powering on the board:
a.) Start with the USB board powered off, and it's power cable disconnected from the
    USB board
b.) Attach the free end of the red wire to a +3.3V power supply. Attach the free end 
    of the black wire to the power supplies return (i.e. signal ground).
c.) Power on the supply.
d.) Plug the power cable in to the USB board.
e.) from a linux bash shell, type: "lsusb". The device "Cypress Semiconductor Corp." 
    with vendor ID number 0x04b4 and product ID number 0x1801 should appear. This is
    the vendor and product ID for the BSD USB interface. After firmware download, the
    product ID changes to 0x1801.

The BSD software package provided by PSU is located in the directory 
BSD_software_Mar20_2013_v1_0.tgz. the Linux "libusb" library is used as the USB API.

To compile the BSD readout software, untar, cd into the directory, and type "make".
This will compile both the SFC software rountines (section 2) and various GSE standalone
programs which demonstrate the operation of the BSD software (section 3).

The USB board uses EP2IN for incoming commands from the SFC, EP4IN as a general 
"USB FIFO reset" command endpoint, EP6OUT for sending responses to these commands, and
EP8OUT for reading out event data. 

Note that in order to trigger the USB board to readout out events, JX1-13 on the 
USB board must be driven high by a +3.3V voltage, and driven low to inhibit triggering.
JX1-13 should ALWAYS be tied to either +3.3V (JX1-49) to generate USB Board event readout
triggers or to 0V (JX1-50) to inhibit these triggers. It can also be driven by a square pulse
with +3.3V to 0V levels. JX1-13 is recommended NOT to be left floating.

A list of commands is available from the document docs/BSD_command_list_Mar20_2013.pdf.

A break down of the event, command, and housekeeping packet format is available from the
document docs/BSD_event_hk_command_packet_Mar20_2013.pdf.

-----------------------------------------------------------------------------------------
2.) SFC software

Overview of the base classes:
BSDUsb: Basic USB endpoint setup and data transfer.
BSDCmd: Class to parse BSD commands and responses from the BSD system.
BSDEvt: Class to parse BSD events. 
HBSD: Encapsulating class for the BSDUsb, BSDCmd, and BSDEvt classes. This is the only
      class the CDAQ should have to interface to directly.

-----------------------------------------------------------------------------------------
3.) GSE standalone programs

The GSE standalone programs show how the HBSD class is used, and will be used for
diagnosising and troubleshooting the BSD on the ground. A description of each program,
along with an example of its usage, will be given below. The source code for these programs 
is located in gse/utils

=================
a.) reset.exe: Load the firware onto the USB board. Note that this command needs to
    be run after each power on of the USB board. 

example: Run the reset program. A message showing the response bytes from EP6 will
	 appear.

=================
b.) EP2OUT.exe: Write N bytes (as hex) to EP2. This is for hardware level testing and
    debugging only

example: Writing the NOP command
$: ./EP2OUT.exe 8f c7 ab ba 00 00 00 00
EP2OUT Write: 0x8F 0xC7 0xAB 0xBA 0x00 0x00 0x00 0x00 

Note: When writing to the USB interface with EP2OUT.exe, one should always follow
up with an EP6IN.exe in order to read the response. Otherwise, commands and responses may
become unsynchronized. The one exception is if the "software generated event"
command is written, in which case the response will be an event data packet on
EP8. In that case, the EP2OUT.exe write should be followed up by a EP8IN.exe read.

=================
c.) EP6IN.exe: Read N bytes (as hex) from EP6. This is for hardware level testing and 
    debugging only.

example: Write the NOP command to EP2 and read out the response from EP6
$: ./EP2OUT.exe 8f c7 ab ba 00 00 00 00
EP2OUT Write: 0x8F 0xC7 0xAB 0xBA 0x00 0x00 0x00 0x00 
$: ./EP6IN.exe 
EP6IN Read:   0x8F 0xC7 0xBC 0xCB 0x00 0x00 0x00 0x01 

=================
d.) EP4OUT.exe: Write N bytes (as hex) to EP4. This is for hardware level testing and
    debugging only. Note that writing ANY packet, no matter the contents, to EP4
    resets the USB FIFOs. This is used to correct sync slips.
$: ./EP4OUT
EP4OUT Write: 0x8F 0xC7 0xAB 0xBA 0x06 0x00 0x00 0x00 

Note: When writing to the USB interface with EP4OUT.exe, one should always follow
up with an EP6IN.exe, to read the response. Otherwise, commands and responses may
become unsynchronized.

=================
e.) EP8IN.exe: Read N bytes (as hex) from EP8. This is for hardware level testing and
    debugging only.

example: First send the command for a software generated event. Then readout EP8.
$: ./EP2OUT.exe 8f c7 ab ba 04 00 00 00
EP2OUT Write: 0x8F 0xC7 0xAB 0xBA 0x04 0x00 0x00 0x00 
$: ./EP8IN.exe 
EP8IN Read:   0x8F 0xC7 0xCD 0xDC 0x04 0x05 0x06 0x07 0x22 0x09 0x0A 0x0B 0x0C 0x0D 0x0E 0x0F 0x10 0x11 0x12 0x13 0x14 0x15 0x16 0x17 0x18 0x19 0x1A 0x1B 0x1C 0x1D 0x1E 0x1F 0x20 0x21 0x22 0x23 0x24 0x25 0x26 0x27 0x28 0x29 0x2A 0x2B 0x2C 0x2D 0x2E 0x2F 0x30 0x31 0x32 0x33 0x34 0x35 0x36 0x37 0x38 0x39 0x3A 0x3B 0x3C 0x3D 0x3E 0x3F 0x40 0x41 0x42 0x43 0x44 0x45 0x46 0x47 0x48 0x49 0x4A 0x4B 0x4C 0x4D 0x4E 0x4F 0x50 0x51

=================
f.) getconf.exe: Get the protoBSD's current configuration (i.e. the value of all BSD parameters and status)

example: Run the command and see the results
$: ./getconf.exe 
main: Opening BSD USB device
tcf_for_bsd.h = 0

               PEDX    D1X_DL    D1X_WD    D2X_DL    D2X_WD    D3X_DL    D3X_WD   DONX_DL   DONX_WD      HV1X      HV2X      HV3X
Quartet1:      2049        19      4999        24      4999       249      4999         4      4999         0         0         0
Quartet2:      2049        19      4999        24      4999       249      4999         4      4999         0         0         0
Quartet3:      2049        19      4999        24      4999       249      4999         4      4999         0         0         0
Quartet4:      2049        19      4999        24      4999       249      4999         4      4999         0         0         0


              PEDTX   D1TX_DL   D2TX_DL   D3TX_DL   D1TX_WD   D2TX_WD   D3TX_WD      THRX      HVTX
Trigger1:      2049         0      4999         4      4999        49      4999      4095         0
Trigger2:      2049         0      4999         4      4999        49      4999      4095         0


            TRIG_MODE      CONF_MODE       COINC_WD       RECOV_WD         DIG_DL     TRIG1_RATE     TRIG2_RATE      SELF_RATE        MT_RATE        T0_RATE
FPGA:               0              0              4            195              8              0              0              0              0              0

            PULSEX_DL    PULSEX_WD   PULSEX_AMP    PULSEX_ON
PULSE1:             0            1         4095            0
PULSE2:            29            1         2867            0

=================
g.) command.exe: Send a pneumonic command to the protoBSD. Pneumonic commands can 
    be "NOP" "GET" "SET" "INIT" and "RESET". See the document
    "BSD_command_list_Mar20_2013.pdf" for a complete list of all possible commands

example: Set a parameter with the SET command. Then readback its value with the GET command. 
$: ./command.exe

Enter commands in mnemonic form, and the 
response will be shown, along with any 
error messages
------------------------------------------------
Send command: 
SET HV22 1000 

Response value: 1000
------------------------------------------------
Send command: 
GET HV22

Response value: 1000
------------------------------------------------
^C

=================
h.) event.exe: Poll for and readout events from the BSD. These are generated by
    +3.3V on JX1-13. Note that this program also creates a timestamped, raw text
    file with the collected BSD data in it.

example: Connect JX1-13 to JX1-49 via the yellow jumper wire. (Note: this provides
+3.3V to JX1-13, hence enabling triggering.) Then run the command:
$: ./event.exe
----------------------------------------------------------------------------------------
Loop 3812
curevt.event_number = 67438087
     EVTNO    EVTCNT     TRIG_MODE     CONF_MODE     TRIG2     TRIG1     EXTIN
  67438087       225             0             1         0         0         1

          PMT1   PMT2   PMT3   PMT4   PMT5   PMT6   PMT7   PMT8   PMT9   PMT10   PMT11   PMT12   PMT13   PMT14   PMT15   PMT16  PMTT1  PMTT2
LG:       2571   3599   4627   5655   6683   7711   8739   9767  10795   11823   12851   13879   14907   15935   16963   17991  19019  20047
HG:       2571   3599   4627   5655   6683   7711   8739   9767  10795   11823   12851   13879   14907   15935   16963   17991  19533  20561
----------------------------------------------------------------------------------------

Many lines similar to the above appear, displaying the contents of the 
BSD event data packet. The EVTCNT parameter gets incremented by the USB board's
firmware. The loop number is simply a firmware counter for the event. All other
data values remain constant.

=================
i.) event_software_generated.exe: Generate a software command and read it out.
    "software generated" means that the event readout is initiated on command,
    not from a hardware trigger. 

example: Connect JX1-13 to JX1-50 via the yellow jumper wire. Note that this means no
hardware trigger is provided to the USB Board. Then run the command:
$: ./event_software_generated.exe
----------------------------------------------------------------------------------------
Loop 3812
curevt.event_number = 67438087
     EVTNO    EVTCNT     TRIG_MODE     CONF_MODE     TRIG2     TRIG1     EXTIN
  67438087       225             0             1         0         0         1

          PMT1   PMT2   PMT3   PMT4   PMT5   PMT6   PMT7   PMT8   PMT9   PMT10   PMT11   PMT12   PMT13   PMT14   PMT15   PMT16  PMTT1  PMTT2
LG:       2571   3599   4627   5655   6683   7711   8739   9767  10795   11823   12851   13879   14907   15935   16963   17991  19019  20047
HG:       2571   3599   4627   5655   6683   7711   8739   9767  10795   11823   12851   13879   14907   15935   16963   17991  19533  20561
----------------------------------------------------------------------------------------




CHANGELOG
---------------------------------------------------------------
Sai Im Tue Apr  2 16:00:36 EDT 2013
./sfc/BSD_defs.h:
Replaced lines:
#define PED1 0x01

with

#define PED1H 0x01

for all PED1 thru PED4

---------------------------------------------------------------
Sai Im Tue Apr  2 16:05:29 EDT 2013
./sfc/BSD_defs.h:
Added lines:

#define PED1L 0x81
#define PED2L 0x82
#define PED3L 0x83
#define PED4L 0x84

Changed line:
#define NGETCONF_PARAMS 129

to

#define NGETCONF_PARAMS 133

---------------------------------------------------------------
Sai Im Tue Apr  2 16:15:59 EDT 2013
.gse/src/command_resolve.cpp:
Changed parameters in char params[Nparams][50]  {...}:
"PED1H","PED2H","PED3H","PED4H"

Added parameters:
"PED1L","PED2L","PED3L","PED4L"

Changed line:
const int Nparams = 129;

to

const int Nparams = 133;

Added numbers in const unsigned char paramval[Nparams]={...}:
129,130,131,132

---------------------------------------------------------------
Sai Im Tue Apr  2 16:30:16 EDT 2013
.sfc/BSDCmd.cpp:
Removed from const unsigned short getconf_params[NGETCONF_PARAMS]:
PED1, PED2, PED3, PED4

Added in to const unsigned short getconf_params[NGETCONF_PARAMS]:
PED1H, PED1L, PED2H, PED2L, PED3H, PED3L, PED4H, PED4L

---------------------------------------------------------------
Sai Im Tue Apr  2 17:18:45 EDT 2013
.gse/util/getconf.cpp:
Replaced in printf("PEDX D1X_DL ...):
PEDX

with

PEDXH   PEDXL   ...

---------------------------------------------------------------
Sai Im Sat Apr  6 10:17:00 EDT 2013
.gse/util/getconf.cpp:
Added after printf("%8d",tcf_for_bsd.data[pos+3]);:     
      pos++;
      printf("%10d",tcf_for_bsd.data[pos+3]);
to make up for adding PEDXL column.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~














---------------------------------------------------------------
Sai Im Thu May  2 13:28:24 EDT 2013
./sfc/.BSD_defs.h:

Added PED5H through PEDT2L at bottom.

Replaced PEDT1 and PEDT2 with PEDT1H and PEDT2H.

Updated #NGETCONF_PARAMS from 133 to 159

---------------------------------------------------------------
Sai Im Thu May  2 13:31:53 EDT 2013
.gse/src/command_resolve.cpp:

xAdded parameters PED5H through PEDT2L in char parms[Nparams][50]

Replaced 

Updated const int Nparams = 159

Added numbers in const unsigned char paramval[Nparams]={...} from 133 to 160.

---------------------------------------------------------------
Sai Im Thu May  2 13:40:40 EDT 2013
.sfc/BSDCmd.cpp:

Added in to const unsigned short getconf_params[NGETCONF_PARAMS]:

PED1H, PED1L, PED2H, PED2L, PED3H, PED3L, PED4H, PED4L
PED5H, etc.
PED9H, etc.
PED13H, etc.

based on quartets.

Added in:

PEDT1H, PEDT1L
and
PEDT2H, PEDT2L

replacing PEDT1 and PEDT2

---------------------------------------------------------------
Sai Im Thu May  2 20:07:02 EDT 2013











~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Sai Im Mon Dec 16 14:08:15 EST 2013
./sfc/BSD_defs.h:

Updated all #define to match with Dec13_2013 command list.

Updated NGETCONF_PARAMS to 144.

---------------------------------------
Sai Im Mon Dec 16 14:11:54 EST 2013
.gse/src/command_resolve.cpp:

Updated char cmds[Ncms][12] with SHUTDOWN command.

Updated all parameters and Nparams = 144 to match command list.

Removed all numbers to match Nparams, updating numbers to match hex value in parameter list.

---------------------------------------
Sai Im Mon Dec 16 14:22:10 EST 2013
.sfc/BSDCmd.cpp:

Updated getconf_params to match command list.

Replaced INIT with SHUTDOWN wherever applicable.

Reorganized getconf_params list to what is best for getconf.exe software.

---------------------------------------
Sai Im Mon Dec 16 15:34:13 EST 2013
../BSD_sim*/BSD_sim*/BSD_simulator_firmware.C

This is where initial values are set up. Had to insert new commands and new parameters for all POR and INIT.

---------------------------------------
Sai Im Mon Dec 16 14:47:23 EST 2013
.gse/util/getconf.cpp:

Reorganized tables in order to fit new parameters adequately.

---------------------------------------
Sai Im Tue Dec 17 14:23:26 EST 2013
BSD-event.h: change #define NUM_RAW_BYTES to 84
EP8IN: had to change nbytes from 82 to 84
BSDsim.ihx EVTSIZE from 82 to 84


----------------------------------------
Sai Im Wed Dec 18 11:36:58 EST 2013
./gse/src/event_resolve.cpp:

curevt.eventcount = bufin[8]
curevt.trig_mode = 0x3F & (bufin[9] >>2) ---> Means that this byte uses 6 of the digits
curevt.conf_mode = 0x07 & (bufin[10] >>5); --->this one uses 3.
curevt.trig2 = 0x01 & (bufin[11] >>2);
curevt.trig1 = 0x01 & (bufin[11] >>1);
curevt.extin = 0x01 & (bufin[11]);  ---> These three are digits 2,1,0 on the same byte (11)

rbc += 8;

---------------------------------------
Sai Im Wed Dec 18 13:59:05 EST 2013
./sfc/BSDCmd.cpp:

Change int BSDCmd::checkEvent() section to have

if( dparam != FPGA )
   return -BSDCMD_CHECK_PARAM_ERR;
return 0;

instead of the other error.

---------------------------------------
Sai Im Wed Dec 18 16:43:36 EST 2013
In the bsd_simulator_firmware:

EVENTs need to output to EP6. Changed by placing EVENT loop within main EP6 output loop.

Remove last few lines that have to do with outputting to EP6 for the RESET_FIFOs.

----------------------------------------

What has ultimately been changed:

Parameters need to be changed and updated: 144 total, but up to 167 hex values. The hex values need to match. PONs have been added at the end.

Commands have been updated such that EVENT works from the command line without error, and RESET_FIFO does not output to EP6. Also need to add SHUTDOWN command in place of INIT and swap SHUTDOWN with EVENT.



sfc/BSD_defs.h
/*
 * Tyler Anderson
 * Wed Jan 23 13:45:04 EST 2013
 * This is the BSD command format
 */
#ifndef _BSD_DEFS_H_
#define _BSD_DEFS_H_

/////////////////////////////////////////////////
// Target defines
/////////////////////////////////////////////////
// b7 is the recoverable error bit. Only set when BSD responds that the
// command has an error.
#define ERROR 0x80
#define ERRORMASK 0x80
#define ERROROFFSET 7

// Action related definitions
#define NOP 0x00
#define GET 0x01
#define SET 0x02
#define EVENT 0x03
#define SHUTDOWN 0x04
#define RESET 0x05 
#define RESET_FIFO 0x06

#define ACTIONMASK 0x07 
#define ACTIONOFFSET 0

// Parameter related definitions
#define FPGA 0x00
#define PED1H 0x01
#define PED2H 0x02
#define PED3H 0x03
#define PED4H 0x04
#define THR1 0x05
#define THR2 0x06
#define D11_DL 0x07
#define D21_DL 0x08
#define D12_DL 0x0A
#define D22_DL 0x0B
#define D13_DL 0x0D
#define D23_DL 0x0E
#define D14_DL 0x10
#define D24_DL 0x11
#define D11_WD 0x13
#define D21_WD 0x14
#define D12_WD 0x16
#define D22_WD 0x17
#define D13_WD 0x19
#define D23_WD 0x1A
#define D14_WD 0x1C
#define D24_WD 0x1D
#define DON1_DL 0x1F
#define DON2_DL 0x20
#define DON3_DL 0x21
#define DON4_DL 0x22
#define DON1_WD 0x23
#define DON2_WD 0x24
#define DON3_WD 0x25
#define DON4_WD 0x26
#define TRIG_MODE 0x27
#define CONF_MODE 0x28
#define COINC_WD 0x29
#define RECOV_WD 0x2A
#define DIG_DL 0x2B
#define PULSE1_DL 0x2C
#define PULSE2_DL 0x2D
#define PULSE1_WD 0x2E
#define PULSE2_WD 0x2F
#define PULSE1_AMP 0x30
#define PULSE2_AMP 0x31
#define PULSE1_ON 0x32
#define PULSE2_ON 0x33
#define HV11 0x34
#define HV21 0x35
#define HV31 0x36
#define HV12 0x37
#define HV22 0x38
#define HV32 0x39
#define HV13 0x3A
#define HV23 0x3B
#define HV33 0x3C
#define HV14 0x3D
#define HV24 0x3E
#define HV34 0x3F
#define HVT1 0x40
#define HVT2 0x41
#define PEDT1H 0x42
#define PEDT2H 0x43
#define D1T1_DL 0x44
#define D2T1_DL 0x45
#define D1T2_DL 0x47
#define D2T2_DL 0x48
#define D1T1_WD 0x4A
#define D2T1_WD 0x4B
#define D1T2_WD 0x4D
#define D2T2_WD 0x4E
#define D11_CM 0x50
#define D21_CM 0x51
#define D12_CM 0x53
#define D22_CM 0x54
#define D13_CM 0x56
#define D23_CM 0x57
#define D14_CM 0x59
#define D24_CM 0x5A
#define D1T1_CM 0x5C
#define D2T1_CM 0x5D
#define D1T2_CM 0x5F
#define D2T2_CM 0x60
#define DON1_CM 0x62
#define DON2_CM 0x63
#define DON3_CM 0x64
#define DON4_CM 0x65
#define D11_ON 0x66
#define D21_ON 0x67
#define D12_ON 0x69
#define D22_ON 0x6A
#define D13_ON 0x6C
#define D23_ON 0x6D
#define D14_ON 0x6F
#define D24_ON 0x70
#define D1T1_ON 0x72
#define D2T1_ON 0x73
#define D1T2_ON 0x75
#define D2T2_ON 0x76
#define DON1_ON 0x78
#define DON2_ON 0x79
#define DON3_ON 0x7A
#define DON4_ON 0x7B
#define TRIG1_RATE 0x7C
#define TRIG2_RATE 0x7D
#define TRIG_SELF_RATE 0x7E
#define TRIG_MT_RATE 0x7F
#define TRIG_T0_RATE 0x80
#define PED1L 0x81
#define PED2L 0x82
#define PED3L 0x83
#define PED4L 0x84
#define PED5H 0x85
#define PED5L 0x86
#define PED6H 0x87
#define PED6L 0x88
#define PED7H 0x89
#define PED7L 0x8A
#define PED8H 0x8B
#define PED8L 0x8C
#define PED9H 0x8D
#define PED9L 0x8E
#define PED10H 0x8F
#define PED10L 0x90
#define PED11H 0x91
#define PED11L 0x92
#define PED12H 0x93
#define PED12L 0x94
#define PED13H 0x95
#define PED13L 0x96
#define PED14H 0x97
#define PED14L 0x98
#define PED15H 0x99
#define PED15L 0x9A
#define PED16H 0x9B
#define PED16L 0x9C
#define PEDT1L 0x9D
#define PEDT2L 0x9E
#define PON_QA 0x9F
#define PON_QB 0xA0
#define PON_QC 0xA1
#define PON_QD 0xA2
#define PON_T1 0xA3
#define PON_T2 0xA4
#define PON_FEEX 0xA5
#define PON_FEEY 0xA6
#define PON_FEEZ 0xA7



#define NGETCONF_PARAMS 144

#endif
